gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsTracer.java

```diff
@@ -42,7 +51,111 @@
 @InternalApi
 public class MetricsTracer implements ApiTracer {
 
-  public MetricsTracer(MethodName methodName, MetricsRecorder metricsRecorder) {}
+  public static final String STATUS_ATTRIBUTE = "status";
+
+  private Stopwatch attemptTimer;
+
+  private final Stopwatch operationTimer = Stopwatch.createStarted();
+
+  private final Map<String, String> attributes = new HashMap<>();
+
+  protected MetricsRecorder metricsRecorder;
+
+  public MetricsTracer(MethodName methodName, MetricsRecorder metricsRecorder) {
+    this.attributes.put("method_name", methodName.toString());
+    this.metricsRecorder = metricsRecorder;
+  }
+
+  @Override
+  public void operationSucceeded() {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.OK.toString());
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void operationCancelled() {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.CANCELLED.toString());
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    attributes.put(STATUS_ATTRIBUTE, extractStatus(error));
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    // no-op
```

Comment:
How is this `attemptStarted(int attemptNumber)` used? If we don't need it, then we can remove this.

Reply:
This is getting called in [CheckingAttemptCallable.java](https://github.com/googleapis/sdk-platform-java/blob/30a366710e6b1afe26df1d1d255fee4abd341575/gax-java/gax/src/main/java/com/google/api/gax/rpc/CheckingAttemptCallable.java#L81). I agree, we do not need it. Removed.

Reply:
I think we will need this, this might be the issue for my test with multiple attempts failing.